src/Orleans.Runtime/Catalog/ActivationData.cs

```diff
@@ -18,1209 +20,1237 @@
+        {
+            limitName = nameof(SiloMessagingOptions.MaxEnqueuedRequestsHardLimit_StatelessWorker);
+            maxRequestsHardLimit = _shared.MessagingOptions.MaxEnqueuedRequestsHardLimit_StatelessWorker;
+            maxRequestsSoftLimit = _shared.MessagingOptions.MaxEnqueuedRequestsSoftLimit_StatelessWorker;
+        }
 
-                return new LimitExceededException(limitName, count, maxRequestsHardLimit, ToString());
-            }
+        if (maxRequestsHardLimit <= 0 && maxRequestsSoftLimit <= 0) return null; // No limits are set
 
-            if (maxRequestsSoftLimit > 0 && count > maxRequestsSoftLimit) // Soft limit
-            {
-                _shared.Logger.LogWarning(
-                    (int)ErrorCode.Catalog_Warn_ActivationTooManyRequests,
-                    "Hot - {Count} enqueued requests for activation {Activation}, exceeding soft limit warning threshold of {SoftLimit}",
-                    count,
-                    this,
-                    maxRequestsSoftLimit);
-                return null;
-            }
+        int count = GetRequestCount();
+
+        if (maxRequestsHardLimit > 0 && count > maxRequestsHardLimit) // Hard limit
+        {
+            _shared.Logger.LogWarning(
+                (int)ErrorCode.Catalog_Reject_ActivationTooManyRequests,
+                "Overload - {Count} enqueued requests for activation {Activation}, exceeding hard limit rejection threshold of {HardLimit}",
+                count,
+                this,
+                maxRequestsHardLimit);
+
+            return new LimitExceededException(limitName, count, maxRequestsHardLimit, ToString());
+        }
 
+        if (maxRequestsSoftLimit > 0 && count > maxRequestsSoftLimit) // Soft limit
+        {
+            _shared.Logger.LogWarning(
+                (int)ErrorCode.Catalog_Warn_ActivationTooManyRequests,
+                "Hot - {Count} enqueued requests for activation {Activation}, exceeding soft limit warning threshold of {SoftLimit}",
+                count,
+                this,
+                maxRequestsSoftLimit);
             return null;
         }
 
-        internal int GetRequestCount()
+        return null;
+    }
+
+    internal int GetRequestCount()
+    {
+        lock (this)
         {
-            lock (this)
-            {
-                return _runningRequests.Count + WaitingCount;
-            }
+            return _runningRequests.Count + WaitingCount;
         }
+    }
 
-        internal List<Message> DequeueAllWaitingRequests()
+    internal List<Message> DequeueAllWaitingRequests()
+    {
+        lock (this)
         {
-            lock (this)
+            var result = new List<Message>(_waitingRequests.Count);
+            foreach (var (message, _) in _waitingRequests)
             {
-                var tmp = _waitingRequests.Select(m => m.Item1).ToList();
-                _waitingRequests.Clear();
-                return tmp;
+                if (message.IsLocalOnly)
```

Comment:
We never forward local-only requests. We could rename this to `IsActivationSpecific` or similar if "local-only" isn't clear enough